sljit_src/sljitNativeARM_64.c

```diff
@@ -2508,19 +2510,58 @@ SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_atomic_load(struct sljit_compiler
 		break;
 	}
 
-	return push_inst(compiler, ins | RN(mem_reg) | RT(dst_reg));
+	return ins;
+}
+
+SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_atomic_load(struct sljit_compiler *compiler, sljit_s32 op,
+	sljit_s32 dst_reg,
+	sljit_s32 mem_reg)
+{
+	CHECK_ERROR();
+	CHECK(check_sljit_emit_atomic_load(compiler, op, dst_reg, mem_reg));
+
+#ifdef __ARM_FEATURE_ATOMICS
+	return sljit_emit_op1(compiler, op, dst_reg, 0, SLJIT_MEM1(mem_reg), 0);
+#else
+	return push_inst(compiler, atomic_load_ins(op) | RN(mem_reg) | RT(dst_reg));
+#endif /* ARM_FEATURE_ATOMICS */
 }
 
 SLJIT_API_FUNC_ATTRIBUTE sljit_s32 sljit_emit_atomic_store(struct sljit_compiler *compiler, sljit_s32 op,
 	sljit_s32 src_reg,
 	sljit_s32 mem_reg,
 	sljit_s32 temp_reg)
 {
-	sljit_ins ins;
+	sljit_ins ins, cmp;
 
 	CHECK_ERROR();
 	CHECK(check_sljit_emit_atomic_store(compiler, op, src_reg, mem_reg, temp_reg));
 
+#if __ARM_FEATURE_ATOMICS
+	cmp = SUBS ^ W_OP;
+	switch (GET_OPCODE(op)) {
+	case SLJIT_MOV32:
+	case SLJIT_MOV_U32:
+		ins = CASL ^ (1 << 30);
+		break;
+	case SLJIT_MOV_U8:
+		ins = CASLB;
+		break;
+	case SLJIT_MOV_U16:
+		ins = CASLH;
+		break;
+	default:
+		ins = CASL;
+		cmp = SUBS;
+		break;
+	}
+	FAIL_IF(push_inst(compiler, atomic_load_ins(op) | RN(mem_reg) | RT(TMP_REG2)));
```

Comment:
This value is in temp_reg. Could be moved to TMP_REGx and do a comparison later.

Reply:
No, that is a fresh atomic load of the memory, which then will let us know if the CAS operation failed by comparing it with the value from temp_reg.

There is a small reservation being done between this and the next instruction, to protect the atomic from being changed by another thread.

Reply:
It looks like the AL variants do the acquire and release as well:

https://developer.arm.com/documentation/ddi0602/2022-06/Base-Instructions/CASB--CASAB--CASALB--CASLB--Compare-and-Swap-byte-in-memory-

That is the normal compare and swap, so perhaps we could use it.

Reply:
Not sure if I understand; the code is already using "acquire/release" semantics in the small critical section.

do you mean to replace `lxdr` with `casa`?, the current version seems to be working fine at least on the M1 (which is the only ARM CPU I have with LSE); it would allow to reintroduce the "fail if temp_reg doesn't match", but thought you were not interested on that "feature".

I have to admit I am still not clear on the use cases for this, so not sure if my tests would give me enough confidence to decide between the two options.

Reply:
> the code is already using "acquire/release" semantics in the small critical section.

except it doesn't work the way I expected, will update once I get one that does and hopefully will be even less code.

Reply:
Something like:
```
TMP_REG1 = tmp_reg
CASALB TMP_REG1, src, [mem]
CMP TMP_REG1, tmp_reg
```
So CASB has 4 varaints, the AL variant do both part of the lock.

Reply:
sadly CASA doesn't work like that; the fetched value is lost so the only way to know if it succeeds or not, is by fetching it again for the CMP as shown in the new version of the code.

I asm using CAS instead as it is "atomic" anyway but left the additional ops defined for swapping them if needed.

Reply:
The original code only used two instructions, so what is the advantage of this variant?
An idea: keep the load the way it is, and just use the correct casxx, that would also reduce the needed instructions.

Reply:
The advantages of this code (compared with LL/SC) are:

* The "critical section" where operations could cause a deadlock (because of bugs, but also "false sharing") is eliminated.
* Performs better under high contention.

The additional instructions are needed to test the result of CAS and can't be eliminated by extending the critical area.  Remember that a small critical area that used a pair of `ldxr`/`casl` was discarded because it didn't work.

Reply:
> sadly CASA doesn't work like that

and I was wrong; the fetched value doesn't get lost and is indeed accessible from `temp_reg` as shown by the latest version of the code